sig-and-wg/sig-core/proposals/common-validator.md

```diff
@@ -0,0 +1,266 @@
+- Different formatting and coding standards in Kyma components.
+- CI validates code in a different environment.
+
+## Solution
+
+Github `kyma-project` organization already contains `makefiles` in every repository and CI pipeline is based on it. In such case, we should unify `makefiles` targets across components/repositories and drop support for any other solution like before-commit, custom validators etc. Thanks to that, whenever contributor will work on `console`, `kyma component` or some other `tool` in our organization, she/he will know how to execute build, test, or format code. It is crucial for development experience.
+
+I will base my proposal on `ui-api-layer` component and targets may not work as it is only an example.
+
+### Makefile Targets
+
+Makefile targets should be simple and should allow contributors to execute all required validations as one command. It should be unified across all `kyma-project` repositories regardless language or type.
+
+Currently, `makefile` in components looks like that:
+
+```makefile
+APP_NAME = ui-api-layer
+IMG = $(DOCKER_PUSH_REPOSITORY)$(DOCKER_PUSH_DIRECTORY)/$(APP_NAME)
+TAG = $(DOCKER_TAG)
+
+.PHONY: ci-pr ci-master ci-release resolve build-and-test build-image push-image
+
+ci-pr: resolve build-and-test build-image push-image
+ci-master: resolve build-and-test build-image push-image
+ci-release: resolve build-and-test build-image push-image
+
+resolve:
+	dep ensure -v -vendor-only
+build-and-test:
+	./before-commit.sh ci
+build-image:
+	docker build -t $(APP_NAME) .
+push-image:
+	docker tag $(APP_NAME) $(IMG):$(TAG)
+	docker push $(IMG):$(TAG)
+```
+
+Only the `ci-*` targets are unified in `kyma-project` organization, because they are required by CI system. Other targets may have different names, because they are not unified and not required by any system. For example `ui-api-layer` has `build-and-test` target, `binding-usage-controller` has `build` - both targets executes `before-commit.sh` script. In such case contributor doesn't know how to work with components in one repository.
+
+As a solution for that I would like to unify targets:
+ - `ci-pr`, `ci-master`, `ci-release` - targets required by CI system
+ - `validate` - execute all validation
+ - `resolve` - download dependencies
+ - `build` - execute build of component
+ - `test` - execute tests for component
+ - `lint` - execute linters for component
+ - `format` - execute code formating for component
+ - `build-image` - build docker image with component
+ - `push-image` - push docker image to repository
+
+ After unification, `makefile` for `ui-api-layer` component will look like that:
+
+ ```makefile
+APP_NAME = ui-api-layer
+IMG = $(DOCKER_PUSH_REPOSITORY)$(DOCKER_PUSH_DIRECTORY)/$(APP_NAME)
+TAG = $(DOCKER_TAG)
+
+.PHONY: ci-pr ci-master ci-release resolve build test format lint build-image push-image
+
+ci-pr: validate build-image push-image
+ci-master: ci-pr
+ci-release: ci-master
+
+validate: resolve build test format lint
```

Comment:
Why do we even need build in here? Isn't test enough?

Reply:
Yes, because we want to validate if the build works ;)